File: sdk/flutter/praxy_flutter/lib/src/realtime/realtime_socket.dart
import 'dart:async';
import 'dart:convert';
import 'dart:math';

import 'package:web_socket_channel/web_socket_channel.dart';

import '../connection_state.dart';

final class _Subscription {
  _Subscription(this.channels, this.controller);
  final List<String> channels;
  final StreamController<Map<String, dynamic>> controller;
}

/// One WebSocket, multiplexed across every [subscribe] caller, reference-counted:
/// connects on the first subscription's first listener, disconnects after a grace
/// period once the last one cancels. Reconnects with exponential backoff *and full
/// jitter* (never Appwrite's lockstep step function — every client reconnecting on
/// the same schedule is its own outage), resending every still-active subscription
/// once the new socket's `connected` frame arrives.
///
/// Constructing more than one of these per client would open two sockets — the
/// exact Appwrite bug this type exists to make structurally impossible; callers
/// must go through `PraxyRealtime`'s single lazily-created instance.
final class RealtimeSocket {
  RealtimeSocket({
    required Uri endpoint,
    required String projectId,
    required Future<String?> Function() ticketProvider,
    Duration idleGracePeriod = const Duration(seconds: 5),
    Duration pingInterval = const Duration(seconds: 20),
    Duration maxBackoff = const Duration(seconds: 30),
  }) : _wsBase = _toWebSocketBase(endpoint, projectId),
       _ticketProvider = ticketProvider,
       _idleGracePeriod = idleGracePeriod,
       _pingInterval = pingInterval,
       _maxBackoff = maxBackoff;

  final Uri _wsBase;
  final Future<String?> Function() _ticketProvider;
  final Duration _idleGracePeriod;
  final Duration _pingInterval;
  final Duration _maxBackoff;
  final Random _random = Random();

  WebSocketChannel? _channel;
  StreamSubscription<dynamic>? _channelSub;
  Timer? _pingTimer;
  Timer? _reconnectTimer;
  Timer? _idleCloseTimer;
  int _reconnectAttempts = 0;
  bool _ready = false;
  bool _disposed = false;
  int _nextSubscriptionId = 0;

  final Map<String, _Subscription> _subscriptions = {};
  final StreamController<PraxyConnectionState> _connectionController = StreamController.broadcast();
  PraxyConnectionState _state = PraxyConnectionState.disconnected;

  /// Replays the current state to every new listener, then forwards live changes —
  /// a plain broadcast stream wouldn't tell a listener who joins mid-connection
  /// anything until the *next* transition.
  Stream<PraxyConnectionState> get connectionStates async* {
    yield _state;
    yield* _connectionController.stream;
  }

  /// Subscribes to [channels] as soon as the returned stream gets a listener, and
  /// unsubscribes when it loses its last one — the unit of ref-counting this class
  /// multiplexes over one socket.
  Stream<Map<String, dynamic>> subscribe(List<String> channels) {
    final id = 'sub${_nextSubscriptionId++}';
    late final StreamController<Map<String, dynamic>> controller;
    controller = StreamController<Map<String, dynamic>>.broadcast(
      onListen: () => _activate(id, channels, controller),
      onCancel: () => _deactivate(id),
    );
    return controller.stream;
  }

  void dispose() {
    _disposed = true;
    _idleCloseTimer?.cancel();
    _reconnectTimer?.cancel();
    _teardownChannel();
    for (final sub in _subscriptions.values) {
      sub.controller.close();
    }
    _subscriptions.clear();
    _setState(PraxyConnectionState.disconnected);
  }

  // ---- activation / ref-counting -----------------------------------------------

  void _activate(String id, List<String> channels, StreamController<Map<String, dynamic>> controller) {
    _idleCloseTimer?.cancel();
    _idleCloseTimer = null;
    _subscriptions[id] = _Subscription(channels, controller);
    _ensureConnecting();
    if (_ready) _sendSubscribe({id: channels});
  }

  void _deactivate(String id) {
    final removed = _subscriptions.remove(id);
    if (removed == null) return;
    if (_ready && _channel != null) _sendUnsubscribe([id]);
    if (_subscriptions.isEmpty) _scheduleIdleClose();
  }

  // ---- connection lifecycle ------------------------------------------------------

  void _ensureConnecting() {
    if (_channel != null || _reconnectTimer != null) return;
    unawaited(_connect());
  }

  Future<void> _connect() async {
    if (_disposed) return;
    _setState(_reconnectAttempts == 0 ? PraxyConnectionState.connecting : PraxyConnectionState.reconnecting);

    final ticket = await _ticketProvider();
    if (_disposed) return;
    var uri = _wsBase;
    if (ticket != null) {
      uri = uri.replace(queryParameters: {...uri.queryParameters, 'ticket': ticket});
    }

    try {
      final channel = WebSocketChannel.connect(uri);
      await channel.ready;
      if (_disposed) {
        await channel.sink.close();
        return;
      }
      _channel = channel;
      _ready = false;
      _channelSub = channel.stream.listen(
        _onMessage,
        onDone: _onSocketDone,
        onError: _onSocketError,
        cancelOnError: true,
      );
    } catch (_) {
      _scheduleReconnect();
    }
  }

  void _onMessage(dynamic raw) {
    final Map<String, dynamic> message;
    try {
      final text = raw is String ? raw : utf8.decode(raw as List<int>);
      message = jsonDecode(text) as Map<String, dynamic>;
    } catch (_) {
      return; // a malformed frame is not this client's problem to crash over
    }

    switch (message['type']) {
      case 'connected':
        _ready = true;
        _reconnectAttempts = 0;
        _setState(PraxyConnectionState.connected);
        _startPing();
        _resubscribeAll();
      case 'event':
        _routeEvent(message['data'] as Map<String, dynamic>?);
      case 'pong':
      case 'response':
      case 'error':
        break; // liveness/acks only; errors have no per-subscription home to route to
      default:
        break; // forward-compatible: an unknown message type is ignored, never thrown on
    }
  }

  void _routeEvent(Map<String, dynamic>? data) {
    if (data == null) return;
    final matched = (data['subscriptions'] as List?)?.cast<String>() ?? const [];
    for (final id in matched) {
      _subscriptions[id]?.controller.add(data);
    }
  }

  void _resubscribeAll() {
    if (_subscriptions.isEmpty) return;
    _sendSubscribe({for (final e in _subscriptions.entries) e.key: e.value.channels});
  }

  void _sendSubscribe(Map<String, List<String>> byId) {
    _send({
      'type': 'subscribe',
      'data': [
        for (final e in byId.entries) {'subscriptionId': e.key, 'channels': e.value},
      ],
    });
  }

  void _sendUnsubscribe(List<String> ids) {
    _send({
      'type': 'unsubscribe',
      'data': [
        for (final id in ids) {'subscriptionId': id},
      ],
    });
  }

  void _send(Map<String, dynamic> message) {
    _channel?.sink.add(jsonEncode(message));
  }

  void _startPing() {
    _pingTimer?.cancel();
    _pingTimer = Timer.periodic(_pingInterval, (_) => _send({'type': 'ping'}));
  }

  void _onSocketDone() => _handleDisconnect();

  void _onSocketError(Object error, StackTrace stackTrace) => _handleDisconnect();

  void _handleDisconnect() {
    _teardownChannel();
    if (_disposed) return;
    if (_subscriptions.isNotEmpty) {
      _scheduleReconnect();
    } else {
      _setState(PraxyConnectionState.disconnected);
    }
  }

  void _teardownChannel() {
    _pingTimer?.cancel();
    _pingTimer = null;
    unawaited(_channelSub?.cancel());
    _channelSub = null;
    _channel = null;
    _ready = false;
  }

  /// Exponential backoff *with full jitter*: `delay = uniform(0, min(maxBackoff, base * 2^attempt))`.
  /// A step function with no jitter (Appwrite's `1s/5s/10s/60s`) puts every
  /// disconnected client back on the wire at the same instant.
  void _scheduleReconnect() {
    _setState(PraxyConnectionState.reconnecting);
    final attempt = _reconnectAttempts++;
    final capMs = min(_maxBackoff.inMilliseconds, 500 * (1 << min(attempt, 10)));
    final delayMs = _random.nextInt(capMs + 1);
    _reconnectTimer?.cancel();
    _reconnectTimer = Timer(Duration(milliseconds: delayMs), () {
      _reconnectTimer = null;
      unawaited(_connect());
    });
  }

  void _scheduleIdleClose() {
    _idleCloseTimer?.cancel();
    _idleCloseTimer = Timer(_idleGracePeriod, () {
      _idleCloseTimer = null;
      if (_subscriptions.isEmpty) _closeSocketOnly();
    });
  }

  void _closeSocketOnly() {
    _reconnectTimer?.cancel();
    _reconnectTimer = null;
    _reconnectAttempts = 0;
    _teardownChannel();
    _setState(PraxyConnectionState.disconnected);
  }

  void _setState(PraxyConnectionState state) {
    _state = state;
    _connectionController.add(state);
  }

  static Uri _toWebSocketBase(Uri endpoint, String projectId) {
    final scheme = endpoint.scheme == 'https' ? 'wss' : 'ws';
    return endpoint.replace(scheme: scheme).resolve('/v1/realtime').replace(
      queryParameters: {'project': projectId},
    );
  }
}

File: sdk/js/packages/core/src/realtime-socket.ts
/**
 * The shared, multiplexed WebSocket connection — one per `RealtimeService`, ref-counted across
 * every `subscribeChannel()` caller. Mirrors `praxy_flutter`'s `RealtimeSocket`
 * (`sdk/flutter/praxy_flutter/lib/src/realtime/realtime_socket.dart`): constructing more than one
 * per client is the exact bug this type exists to make structurally impossible.
 *
 * Uses the global `WebSocket` (available in browsers, edge runtimes, and Node 22+) — no
 * `node:*` import, so this stays safe to pull into `praxyMiddleware()`'s Edge Runtime bundle.
 */

export type ConnectionState = "disconnected" | "connecting" | "connected" | "reconnecting";

export interface ServerFrame {
  type: string;
  data?: unknown;
}

export interface Unsubscribe {
  unsubscribe(): void;
}

export interface RealtimeSocketConfig {
  /** `wss://host/v1/realtime?project=<id>` — a ticket is appended as `&ticket=...` per connect attempt. */
  wsUrl: string;
  /** Mints a fresh single-use ticket; `null` connects as a guest (no session/JWT to mint from). */
  mintTicket: () => Promise<string | null>;
}

const PING_INTERVAL_MS = 20_000;
const IDLE_CLOSE_MS = 5_000;
const BASE_BACKOFF_MS = 500;
const MAX_BACKOFF_MS = 30_000;
const MAX_BACKOFF_ATTEMPT = 10;

export class RealtimeSocket {
  private ws: WebSocket | null = null;
  private state: ConnectionState = "disconnected";
  private attempt = 0;
  private explicitlyClosed = false;

  private reconnectTimer: ReturnType<typeof setTimeout> | null = null;
  private pingTimer: ReturnType<typeof setInterval> | null = null;
  private idleTimer: ReturnType<typeof setTimeout> | null = null;

  private readonly channelListeners = new Map<string, Set<(frame: ServerFrame) => void>>();
  private readonly subscriptionIdByChannel = new Map<string, string>();
  private readonly stateListeners = new Set<(state: ConnectionState) => void>();
  private nextSubscriptionId = 1;

  constructor(private readonly config: RealtimeSocketConfig) {}

  subscribeChannel(channel: string, listener: (frame: ServerFrame) => void): Unsubscribe {
    this.explicitlyClosed = false;
    this.cancelIdleClose();

    let listeners = this.channelListeners.get(channel);
    const isNewChannel = !listeners;
    if (!listeners) {
      listeners = new Set();
      this.channelListeners.set(channel, listeners);
    }
    listeners.add(listener);

    if (isNewChannel && this.state === "connected") this.sendSubscribe(channel);
    this.ensureConnecting();

    return {
      unsubscribe: () => {
        listeners!.delete(listener);
        if (listeners!.size === 0) {
          this.channelListeners.delete(channel);
          if (this.state === "connected") this.sendUnsubscribe(channel);
          else this.subscriptionIdByChannel.delete(channel);
          this.scheduleIdleCloseIfEmpty();
        }
      },
    };
  }

  subscribeConnectionState(listener: (state: ConnectionState) => void): Unsubscribe {
    this.stateListeners.add(listener);
    listener(this.state);
    return { unsubscribe: () => this.stateListeners.delete(listener) };
  }

  close(): void {
    this.explicitlyClosed = true;
    this.teardown();
    this.setState("disconnected");
  }

  private ensureConnecting(): void {
    if (this.ws || this.state === "connecting" || this.state === "reconnecting") return;
    this.attempt = 0;
    void this.connect();
  }

  private async connect(): Promise<void> {
    this.setState(this.attempt > 0 ? "reconnecting" : "connecting");

    const ticket = await this.config.mintTicket().catch(() => null);
    const url = ticket ? `${this.config.wsUrl}&ticket=${encodeURIComponent(ticket)}` : this.config.wsUrl;

    let socket: WebSocket;
    try {
      socket = new WebSocket(url);
    } catch {
      this.scheduleReconnect();
      return;
    }
    this.ws = socket;

    socket.addEventListener("message", (event) => this.handleMessage(String((event as MessageEvent).data)));
    socket.addEventListener("close", () => this.handleClose());
    socket.addEventListener("error", () => {
      /* the subsequent close event drives reconnect — nothing extra to do here */
    });
  }

  private handleMessage(raw: string): void {
    let frame: ServerFrame;
    try {
      frame = JSON.parse(raw) as ServerFrame;
    } catch {
      return;
    }

    if (frame.type === "connected") {
      this.attempt = 0;
      this.setState("connected");
      this.resubscribeAll();
      this.startPing();
      return;
    }
    if (frame.type === "pong") return;
    if (frame.type === "event") {
      const channels = (frame.data as { channels?: string[] } | undefined)?.channels ?? [];
      for (const channel of channels) {
        this.channelListeners.get(channel)?.forEach((listener) => listener(frame));
      }
    }
  }

  private handleClose(): void {
    this.ws = null;
    this.stopPing();
    this.subscriptionIdByChannel.clear();
    if (this.explicitlyClosed) {
      this.setState("disconnected");
      return;
    }
    this.setState("disconnected");
    if (this.channelListeners.size > 0) this.scheduleReconnect();
  }

  /** Exponential backoff with full jitter — avoids every disconnected client retrying in lockstep. */
  private scheduleReconnect(): void {
    if (this.explicitlyClosed) return;
    const cappedAttempt = Math.min(this.attempt, MAX_BACKOFF_ATTEMPT);
    const cap = Math.min(MAX_BACKOFF_MS, BASE_BACKOFF_MS * 2 ** cappedAttempt);
    const delay = Math.random() * cap;
    this.attempt += 1;
    this.reconnectTimer = setTimeout(() => {
      this.reconnectTimer = null;
      if (this.channelListeners.size > 0) void this.connect();
    }, delay);
  }

  private resubscribeAll(): void {
    for (const channel of this.channelListeners.keys()) this.sendSubscribe(channel);
  }

  private sendSubscribe(channel: string): void {
    const subscriptionId = `s${this.nextSubscriptionId++}`;
    this.subscriptionIdByChannel.set(channel, subscriptionId);
    this.send({ type: "subscribe", data: [{ subscriptionId, channels: [channel] }] });
  }

  private sendUnsubscribe(channel: string): void {
    const subscriptionId = this.subscriptionIdByChannel.get(channel);
    this.subscriptionIdByChannel.delete(channel);
    if (subscriptionId) this.send({ type: "unsubscribe", data: [{ subscriptionId }] });
  }

  private send(message: unknown): void {
    if (this.ws?.readyState === WebSocket.OPEN) this.ws.send(JSON.stringify(message));
  }

  private startPing(): void {
    this.stopPing();
    this.pingTimer = setInterval(() => this.send({ type: "ping" }), PING_INTERVAL_MS);
  }

  private stopPing(): void {
    if (this.pingTimer) clearInterval(this.pingTimer);
    this.pingTimer = null;
  }

  /** Closes the socket once the last subscription cancels, after a short grace period — avoids
   *  thrashing the connection across React's rapid mount/unmount cycles (e.g. StrictMode). */
  private scheduleIdleCloseIfEmpty(): void {
    if (this.channelListeners.size > 0) return;
    this.cancelIdleClose();
    this.idleTimer = setTimeout(() => {
      if (this.channelListeners.size === 0) this.teardown();
    }, IDLE_CLOSE_MS);
  }

  private cancelIdleClose(): void {
    if (this.idleTimer) clearTimeout(this.idleTimer);
    this.idleTimer = null;
  }

  private teardown(): void {
    this.cancelIdleClose();
    if (this.reconnectTimer) clearTimeout(this.reconnectTimer);
    this.reconnectTimer = null;
    this.stopPing();
    this.ws?.close();
    this.ws = null;
  }

  private setState(state: ConnectionState): void {
    if (this.state === state) return;
    this.state = state;
    this.stateListeners.forEach((listener) => listener(state));
  }
}
